File: vepserver/src/main/scala/vep/model/theater/Theater.scala
package vep.model.theater

import spray.json.{JsArray, JsValue, RootJsonFormat}
import vep.model.JsonImplicits
import vep.model.common.{ErrorCodes, VerifiableMultiple}
import vep.utils.StringUtils

case class Theater(id: Long, canonical: String, name: String, address: String, content: Option[String], fixed: Boolean, plan: Option[String], maxSeats: Option[Int])

case class TheaterFormBody(name: String, address: String, content: Option[String], fixed: Boolean, plan: Option[String], maxSeats: Option[Int])

case class TheaterForm(canonical: String, name: String, address: String, content: Option[String], fixed: Boolean, plan: Option[String], maxSeats: Option[Int]) extends VerifiableMultiple {
  override def verify: Boolean = {
    if (StringUtils.isBlank(canonical)) {
      addError("canonical", ErrorCodes.emptyField)
    } else if (StringUtils.isNotCanonical(canonical)) {
      addError("canonical", ErrorCodes.invalidCanonical)
    } else if (canonical.length > 255) {
      addError("canonical", ErrorCodes.bigString)
    }
    if (StringUtils.isBlank(name)) {
      addError("name", ErrorCodes.emptyField)
    } else if (name.length > 255) {
      addError("name", ErrorCodes.bigString)
    }
    if (StringUtils.isBlank(address)) {
      addError("address", ErrorCodes.emptyField)
    }
    if (fixed) {
      if (!plan.exists(StringUtils.isNotEmpty)) {
        addError("plan", ErrorCodes.emptyField)
      } else if (!plan.exists(StringUtils.isJson)) {
        addError("plan", ErrorCodes.invalidJson)
      }
    } else {
      if (maxSeats.isEmpty) {
        addError("maxSeats", ErrorCodes.emptyField)
      } else if (maxSeats.get <= 0) {
        addError("maxSeats", ErrorCodes.negativeOrNull)
      }
    }

    hasNotErrors
  }
}

case class TheaterSeq(theaters: Seq[Theater])

object TheaterForm {
  def fromTheaterFormBody(canonical: String, theaterFormBody: TheaterFormBody) = TheaterForm(canonical, theaterFormBody.name, theaterFormBody.address, theaterFormBody.content, theaterFormBody.fixed, theaterFormBody.plan, theaterFormBody.maxSeats)
}

object TheaterImplicits extends JsonImplicits {
  implicit val impTheater: RootJsonFormat[Theater] = jsonFormat(Theater.apply, "id", "canonical", "name", "address", "content", "fixed", "plan", "maxSeats")
  implicit val impTheaterFormBody: RootJsonFormat[TheaterFormBody] = jsonFormat(TheaterFormBody.apply, "name", "address", "content", "fixed", "plan", "maxSeats")

  implicit val impTheaterSeq = new RootJsonFormat[TheaterSeq] {
    override def read(value: JsValue) = TheaterSeq(value.convertTo[Seq[Theater]])

    override def write(f: TheaterSeq) = JsArray(f.theaters map { theater => impTheater.write(theater) }: _*)
  }
}

File: vepweb/lib/model/theater/Theater.dart
part of vep.model.theater;

class Theater {
  String canonical;
  String name;
  String address;
  String content;
  bool fixed;
  String plan;
  int maxSeats;

  TheaterForm toTheaterForm() {
    var theaterForm = new TheaterForm();
    theaterForm.canonical = canonical;
    theaterForm.name = name;
    theaterForm.address = address;
    theaterForm.content = content;
    theaterForm.fixed = fixed;
    theaterForm.plan = plan;
    theaterForm.maxSeats = maxSeats;
    return theaterForm;
  }
}

class TheaterForm {
  @jsonIgnore
  String canonical;
  String name;
  String address;
  String content;
  bool fixed;
  String plan;
  int maxSeats;
}

class Seat {
  /// Code
  String c;

  /// Type
  String t;

  /// x-axis
  double x;

  /// y-axis
  double y;

  /// Width
  double w;

  /// Height
  double h;

  @jsonIgnore
  bool selected;

  @jsonIgnore
  String get className => 'theater-plan-seat-' + (selected != null && selected ? 'selected' : (t != null ? t : 'normal'));
}